crates/parcel_plugin_transformer_js/src/transformer.rs

```diff
@@ -1,24 +1,947 @@
-  pub fn new(_ctx: &PluginContext) -> Self {
+impl ParcelJsTransformerPlugin {
+  pub fn new() -> Self {
     Self {}
   }
 }
 
-impl TransformerPlugin for ParcelTransformerJs {
+impl TransformerPlugin for ParcelJsTransformerPlugin {
+  /// This does a lot of equivalent work to `JSTransformer::transform` in
+  /// `packages/transformers/js`
   fn transform(
     &mut self,
-    _context: &mut RunTransformContext,
-    _asset: TransformationInput,
-  ) -> Result<TransformResult, anyhow::Error> {
-    todo!()
+    context: &mut RunTransformContext,
+    input: TransformationInput,
+  ) -> Result<TransformResult, Error> {
+    let file_system = context.file_system();
+    let source_code = input.read_source_code(file_system)?;
+
+    let transformation_result = parcel_js_swc_core::transform(
+      parcel_js_swc_core::Config {
+        filename: input
+          .file_path()
+          .to_str()
+          .ok_or_else(|| anyhow!("Invalid non UTF-8 file-path"))?
+          .to_string(),
+        code: source_code.bytes().to_vec(),
+        source_type: parcel_js_swc_core::SourceType::Module,
+        ..parcel_js_swc_core::Config::default()
+      },
+      None,
+    )?;
+
+    // TODO handle errors properly
+    if let Some(errors) = transformation_result.diagnostics {
+      return Err(anyhow!(format!("{:#?}", errors)));
+    }
+
+    let asset = Asset::new_empty(input.file_path().to_path_buf(), source_code);
+    let config = parcel_js_swc_core::Config::default();
+    let options = ParcelOptions::default();
+    let result = convert_result(asset, &config, transformation_result, &options)
+      // TODO handle errors properly
+      .map_err(|_err| anyhow!("Failed to transform"))?;
+
+    Ok(result)
+  }
+}
+
+fn convert_result(
+  mut asset: Asset,
+  transformer_config: &parcel_js_swc_core::Config,
+  result: parcel_js_swc_core::TransformResult,
+  options: &ParcelOptions,
+) -> Result<TransformResult, Vec<Diagnostic>> {
+  let asset_file_path = asset.file_path.to_path_buf();
+  let asset_environment = asset.env.clone();
+  let asset_id = asset.id();
+
+  if let Some(shebang) = result.shebang {
+    asset.set_interpreter(shebang);
+  }
+
+  let (mut dependency_by_specifier, invalidate_on_file_change) = convert_dependencies(
+    transformer_config,
+    result.dependencies,
+    &asset_file_path,
+    &asset_environment,
+    asset_id,
+  )?;
+
+  if result.needs_esm_helpers {
+    let mut dependency_flags = DependencyFlags::empty();
+    dependency_flags.set(
+      DependencyFlags::HAS_SYMBOLS,
+      result.hoist_result.is_some() || result.symbol_result.is_some(),
+    );
+    let dependency = make_esm_helpers_dependency(
+      options,
+      &asset_file_path,
+      asset_environment,
+      dependency_flags,
+      asset_id,
+    );
+    dependency_by_specifier.insert(dependency.specifier.as_str().into(), dependency);
+  }
+
+  if let Some(hoist_result) = result.hoist_result {
+    asset.flags.set(AssetFlags::HAS_SYMBOLS, true);
```

Comment:
Is `has_symbols` needed? I find it weird, why can't we check for the `symbols` len?

Reply:
There is actually a difference between a module having no exports and an asset not having analyzed the symbols at all. In the latter case, we must assume that any import to that module exists rather than throwing a "does not export" error.

What we should probably do is make `asset.symbols` an `Option<Vec<Symbol>>` instead. This flag is a holdover from the old `db2` branch.

Reply:
I've documented but won't fix in this diff